### DISCUSSION TITLE
Add service to retrieve schedule configuration

### DISCUSSION THREAD

**Discussion by rikroe** [home-assistant/architecture#1115] [Entity Models] [answered]
## Proposal

I propose to add a new service to the `schedule` helper entity that returns the full weekly schedule configured by the user (via YAML or the UI). Currently, the schedule entity works like a binary sensor (on/off state) and has the next event (i.e. next time switching between on and off).

The response of this service can be used by automations (with some templating) to convert it into a format which can be used as input for other services pushing this schedule to an actual device.
It is up to the user's discretion to handle the resulting output via automation/blueprint etc.

This is a similar approach to deal with complex data structures (i.e. non-state) fields as discussed for `calendar.get_events` and `weather.get_forecasts` in https://github.com/home-assistant/architecture/discussions/948.

## Context

Home Assistant for many users is the central place to view, interact and **configure** their smart home gadgets. Some devices (in my case heating-related, i.e. the central heating and radiator TRVs) are "smart" enough to have their own schedule to operate on their own, configured via Bluetooth/WiFi and the manufacturer's app.

These device may have integrations or custom components allowing users to interact with the device (i.e. setting the current temperature to stay in the climate domain). However, most of these "smart" devices will override any manual value configured Home Assistant when the next schedule change is planned.

Being able to override the device's schedule based on information available in Home Assistant would enable users to:
- not use outdated, hard to use manufacturer apps to control schedules
- more automation options such as using different schedules based on other entities (e.g. different schedules based on device/person location)

## Reference implementation

https://github.com/home-assistant/core/pull/121904

**Comment by allenporter** [comment]
One naive idea: Could schedule also create a calendar entity? This would give it services for exporting the schedule for any arbitrary date/time. This could let users visualize the schedule on the calendar UI, however, this could also be confusing since the roles of schedules and calendars are similar but different.  Implementation wise, it would work similar to `local_calendar` with an event and rrule for each entry.

**Reply by rikroe** [reply]
Thanks for your answer! Much appreciate it for bringing in another idea!

The schedule entity (in contrast to the local_calendar) will simply turn on or off based on the schedule. Very simple and can be automated easily with a state trigger. The calendar integration has its own set of triggers, making it much more flexible but also a little more complex. 
I would assume that if the flexibility of triggers or other than weekly schedules is required, users already are using a calendar-based integration.
However, adding/syncing schedules to the calendar would in my opinion make it more confusing (you mentioned it, what is the leading integration, ...) so I would not do that.

Today, I could build my automation using a calendar, however I would always need to make sure to select a specific week (in the past or with even more scripting the current week) to get a schedule I can write to my devices. Again, not the simplest but somehow possible.

While I very much appreciate your idea, I don't fully understand why the addition of a "read-only" service is that "controversial".

**Comment by pebblebed-tech** [comment]
I too have been looking at this in respect to providing on the device scheduling to ESPHome. What I proposing to implement is a HACS card that gets the schedule via the callws method that the scheduler provides and then the card can call a service implemented by the device to transfer the schedule. See https://discord.com/channels/330944238910963714/1257019582112334014/threads/1328377096137085091 for discussion

**Comment by abmantis** [comment, marked as answer]
We discussed this on the core team meeting and the proposal was approved.
I've reopened and unlocked the reference PR.

Thanks!

**Reply by rikroe** [reply]
Thanks!
